### ISSUE TITLE
liveness probe /health/live not UP while DB migrations initialization

### ISSUE THREAD

**Issue by dosmanak** [keycloak/keycloak#35261]
### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

core

### Describe the bug

During fresh intitialization for keycloak with mysql DB, quite a lot migrations are invoked and the keycloak liveness probe does not serve. The pod is restarted due to liveness failure and the subsequent restarted container fails due database inconsistence.

### Version

26.0.6

### Regression

- [ ] The issue is a regression

### Expected behavior

The liveness probe should be UP while db migration are in progress

### Actual behavior

The failed liveness proble causes container restart

### How to Reproduce?

Give the pod little CPU resources and set liveness probe /health/live with startupDelay e.g. 60s.

### Anything else?

Thank you for great work!

**Comment by sschu** [comment]
Is this about the Keycloak operator?

**Comment by shawkins** [comment]
Confirmed there is an issue here. The startup probe / liveness probes are not responsive while a migration is running - a 503 is shown instead. The default is to have a 10 minute startup probe (1 second period 600 failures) - if the migration exceed that time, then it will be restarted.

It's a related, but separate issue, that the db is left in an inconsistent state and the migration logic is unable to resume. That is not expected, but I believe that has been seen before. @dosmanak can you provide what errors you see on subsequent startups / migration attempts? cc @ahus1 

@dosmanak as a workaround, you may override the default startup probe with a timeout that is long enough to account for your migration.

To resolve this we have a couple of options for the operator:
- increase the default startup time as a quick fix
- ensure that the startup / liveness probes succeed, but readiness does not. Not sure about the scope of this yet as it doesn't appear that vertx is fully initialized yet when the migration is running.

cc @vmuzikar @mabartos @Pepo48

**Comment by vmuzikar** [comment]
This indeed sounds serious.

> increase the default startup time as a quick fix

I'd rather document how to increase the time manually using pod template. We do not know a "safe timeout" to cover longer migrations. And it'd be a breaking change if we wanted to decrease it back to the original value once we have a proper fix.

> ensure that the startup / liveness probes succeed, but readiness does not. Not sure about the scope of this yet as it doesn't appear that vertx is fully initialized yet when the migration is running.

Yes, this sounds good but as it will be probably more involved I'd suggest doing the quick fix (docs update) first.

**Comment by ahus1** [comment]
@shawkins - that's a great investigation, thank you for that. As a first step, I'd say extending the default timeout of the startup probe would be a good remedy. 

Even if we get the startup probe working while the database migration is under way, it might be a challenge to see that the liveness probe returns success and the readiness probe returns false. So the extended timeout of the startup probe (in the area of let's say 60 minutes) would be a default that would stay with us for a while AFAIK. 

There might be another timeout for migrations: The database transaction timeout is set to 300 seconds (`quarkus.transaction-manager.default-transaction-timeout`). I didn't test it yet, but I suspect any database schema migration might be constrained by that timeout as well - either on a per-change or per-migration. It would be good if the core team together with the CND team could verify if there is this second challenge. Maybe it would be good to have a different transaction timeout set for database schema migrations. We had a discussion pointing in that direction in #35290

cc: @sguilhen, @pedroigor

**Comment by dosmanak** [comment]
Thank you for confirmation. I do not have the test lab ready for new keycloak initialization as we had to proceed with the project.
Currently there is only probe startup delay set in out manifests so I will add a startup probe.

Still I think the liveness probe should be UP while migrations are in progress and the readiness probe should be UP in the moment the migrations are done.

**Comment by sschu** [comment]
@ahus1 How about setting the startup probe timeout in sync with the transaction timeout by default? In my experience, there are two factors causing a slow startup: a) lack of available CPU time b) long-running database migrations. This means, if there are migrations, the startup would always fail after the transaction timeout. The only case where a longer startup probe would make sense is when there are no migrations and the only cause for the slow startup is lack of CPU. And if Keycloak needs 5 minutes to start up without any migrations, you should really rather give more CPU.

**Comment by ahus1** [comment]
@sschu - let's see first if and how the order different probes getting ready and the DB migration can be changed at all. I hope that @shawkins can provide insights here. 

In the meantime, the only option I see to by default extend the startup time to cover startup and the migration (which can be in the are of tens of minutes, maybe even an hour).

**Comment by sschu** [comment]
But only if you also increase the transaction timeout to that value, otherwise the migration will fail after 5 minutes. Or do you assume it is 5 minutes per individual changeset in the migration?

**Comment by ahus1** [comment]
> But only if you also increase the transaction timeout to that value, otherwise the migration will fail after 5 minutes. Or do you assume it is 5 minutes per individual changeset in the migration?

That will also need investigation - maybe we can have a different timeout during the migrations, maybe not. 

Another option is to run database migrations in an init-container, and it would give more flexibility on startup-probes and different timeout configurations. I'd say this would be a plan B if the re-ordering doesn't work. The downside would be that the init container would start every time before the main container, so it would add to the startup time.

**Comment by shawkins** [comment]
@ahus1 @sschu @dosmanak are you good with spawning a separate issue to cover whatever is going wrong with the migration after the termination? If that were addressed - any changes to the probes would simply be nice to have, rather than required. It would also in general improve the stability of migrations even when not on openshift.

**Comment by sschu** [comment]
@shawkins I don't really think probe changes would be just nice to have. Even if you make the migrations idempotent so they can be rerun, they still need enough time to finish. Maybe having a really long timeout for the startup probe as suggested by @ahus1 is the best solution to the problem reported here and the problem to allow rerun of migrations can be handled in a separate issue.

**Comment by shawkins** [comment]
> @shawkins I don't really think probe changes would be just nice to have. Even if you make the migrations idempotent so they can be rerun, they still need enough time to finish.

This is true iff the migration can only be run as a whole, which seems fragile given how long this seems to take. I'll test some more to see if this seems to be the case. Even if we adjust the probes in this case there will still be an issue that an interruptted migration left the database in an indeterminant state.

**Comment by sschu** [comment]
It could also be just one step the migration that takes long and would never finish without an increased timeout. Thinking of changes on the USER_ATTRIBUTES for example where there might be plenty of.

**Comment by vmuzikar** [comment]
Increasing the startup probe timeout is really not a universal solution. We can't guess how long does the migration take or what are the transaction timeouts. IMHO a proper solution is to fix the migration behaviour to ensure if a migration is interrupted (by the probes or anything else – it's not specific to K8s environment), it can continue the next it it's ran. This also expects we use a reasonably large transactions, rather smaller than bigger. Not sure though what are our technical limitations coming from Liquibase.

Hence I also see increasing the probes timeout as something nice to have.

**Comment by sschu** [comment]
If you have a migration with a really long-lasting step higher than the startup probe timeout, it will never finish no matter how often you restart. I am talking about things that run in a transaction and are automatically rolled back i.e. data, not schema changes. I also don't really see a downside to a long startup probe. If it is too short, you might end up in a restart loop if the underlying problem does not go away on its own. If it is too long, you might wait too long before trying to restart to fix a problem that is just fixed by trying again - I wonder what would that be?

**Comment by shawkins** [comment]
Confirmed VertxHttpProcessor.openSocket is dependent on ApplicationStartBuildItem, which makes it wait for all startup events before proceeding. 

There are 3 courses of action mentioned above:

1. Simply increase the startup probe time - but to what value? 30 minutes, 1 hour? It is still a valid concern what is happening when long-running migrations are interrupted. It was never posted here, so it's not clear why the user's database was in an inconsistent state.

2. If we wanted to reorder things, that would involve steps like:
- defer the startup logic related to migration, but probably only for containerized deployment
- add a specific startup HealthCheck, which will be updated when the migration task finishes
- however we still have to increase the startup probe time. There doesn't seem to be much advantage in doing this approach as you can't make the startup probe adaptable to a third (wait longer / still migrating) state - you only get to deal with up or down. About the only thing you gain is the potential to expose more information though in the probe response.

3. Use an init container. We'd probably have to introduce some new command or arg for this such as "start --migrate-only" or just "migrate". The operator would configure basically the exactly the same process / command as the regular start but with the additional bit that would make the process exit after startup / migration completes - like the other non-server commands. This has a moderate amount of work between the operator and picocli changes, and the concern over greater startup times is valid especially because not everyone is even using optimized operator images.

@vmuzikar @sschu @ahus1 @dosmanak based upon the above, I'm fine with proceeding with just increasing the startup probe time. Are there any qualms with 1 hour?

**Comment by ahus1** [comment]
@shawkins - I'm +1 for increasing the startup probe time, and 1 hour might be a good fit here. This could be a quick fix for the symptoms we see, and it could even be backported to KC 26

I might have a twist for the option 2: If we don't implement a startup health check, but a readiness health check, we could come up with the following: 

* Keycloak initializes, and the startup probe is now "up" together with the liveness probe that is "up". The readiness probe is still "down", so no requests would be forwarded to Keycloak.
* Keycloak performs the migration. Once if that is complete, the readiness would show "up" as well, and Keycloak would serve requests. 

This would then allow a short startup probe phase again. The liveness probe would watch over the migration. AFAIK there is no time limitation in Kubernetes for the readiness probe to be "down".

WDYT?

**Comment by slaskawi** [comment]
From my experience the primary reason for using the Startup Probe is applying different checking intervals and different criteria for applying Pod's `restartPolicy`. 

Just to illustrate this with an example - when Keycloak serves traffic without interruptions, we want to kill it after 30 seconds (3 failures, 10 seconds each). During the migration, we tolerate (via the Startup Probe) up to 600 seconds (600 failures, 1 second each). It goes without saying that in all cases the Readiness Probe should be down - we don't want any traffic to be routed to an unhealthy, not fully-operational Keycloak instance.

In my opinion, we could try to calculate the worst case scenario for the database migration and add some healthy offset to it. I might be misinterpreting the code but the migrations seem to be running in a transaction so the absolute transaction timeout will be our limiting factor: https://github.com/keycloak/keycloak/blob/99478887a47796ecd65ae7347c9ffd44fe7f6755/model/jpa/src/main/java/org/keycloak/connections/jpa/DefaultJpaConnectionProviderFactory.java#L335-L347

If my understanding is correct (which has also been suggested in https://github.com/keycloak/keycloak/issues/35261#issuecomment-2519610858), the default timeout for migrations is 300 seconds. If migrations fail with a timeout, we need to wait additional 300 seconds for the Startup Probe failures until Kubernetes restarts the Pod. 

A while back I saw a suggested feature, https://github.com/keycloak/keycloak/issues/26889, that would enable separating transaction timeouts based on categories - one of which are migrations. I believe we could achieve the best user experience if we implemented that feature first, set the migrations timeout to some very high value value like 1 hour and then use the same timeout also for the Startup Probe. In addition to that, we might want to consider crashing the Pod if migrations are not successful as waiting for the Startup Probe failures might not always be optimal.

**Comment by vmuzikar** [comment]
From my perspective there are several layers in this problem.

The first one is the probes. I'm 100 % in favor of reordering the migration logic to leverage a readiness check instead of startup probe. IMHO it is a more correct behavior to fail a readiness check instead of startup/liveness check during migration, as migration timeout is not really something that should cause a pod restart. That said, I'm not sure if we can really achieve that due to:

> Confirmed VertxHttpProcessor.openSocket is dependent on ApplicationStartBuildItem, which makes it wait for all startup events before proceeding.

However there's alway a "hard" limit of transaction timeouts, so in any case there will be a timeout that fails the migration, regardless of the probes. So in the end it doesn't really matter if we keep using the startup probe as we need a timeout in any case. Now this:

> we could try to calculate the worst case scenario for the database migration and add some healthy offset to it

might be a bit tricky as there might be vast differences in the time required for migration depending on many factors.

Thinking about it more now, IMHO we need to introduce a new config option, something like `migration-timeout` (or possibly just `transaction-timeout-migration`). It would:
* set transaction timeouts for migrations;
* be reflected by the startup probe;
* made sure the migration is atomic as a whole and if it timeouts, it doesn't leave the DB in an inconsistent state.

**Comment by shawkins** [comment]
> @shawkins - I'm +1 for increasing the startup probe time, and 1 hour might be a good fit here. This could be a quick fix for the symptoms we see, and it could even be backported to KC 26

Ok, let's start with that.

> This would then allow a short startup probe phase again. The liveness probe would watch over the migration. AFAIK there is no time limitation in Kubernetes for the readiness probe to be "down".

That is true. I was trying to keep the long running operation associated with the startup, but having it associated with readiness would work better. The level of effort remains about the same, but we may want to adjust our readiness probe period / failure threshold - the current 10 second period might be too long if we are more commonily going to be unready.

I'll create a separate pr for increasing the startup timeout so that this issue can track this change.

> Thinking about it more now, IMHO we need to introduce a new config option, something like migration-timeout (or possibly just transaction-timeout-migration). It would:

Agreed and this also addresses @slaskawi 's concerns. But let's create another issue(s) for this - what @ahus1 mentions in https://github.com/keycloak/keycloak/issues/35261#issuecomment-2519610858 will affect how this is documented / implemented. Once the core team confirms what property or properties will be used to configure timeouts, then we can incorporate that into the operator / CR. If we do the reordering of the startup logic in the meantime, then we won't have to futher adjust the startup probe for this.

> made sure the migration is atomic as a whole and if it timeouts, it doesn't leave the DB in an inconsistent state.

It would be great if @dosmanak could provide what state things ended up in. Should we proactively create an issue to investigate it further, or wait until we have concrete details?

**Comment by vmuzikar** [comment]
> decouples things a bit

That's what I'm not sure would be too beneficial in the end (despite my previous comments from Dec). I agree handling the migration check in readiness endpoint is more correct than startup/liveness. But we're gonna hit the transaction timeout anyway. So I see the logic reordering only as something nice to have as what we truly need to handle is a configurable "migration timeout".

**Comment by shawkins** [comment]
> But we're gonna hit the transaction timeout anyway.

Is it clear to you from @ahus1 comments how this currently or should work? Will the timeout eventually be for each migration step, or for the whole migration? If it's per step, how will know how many steps are going to be run? Regardless how would a user know, other than trial and error, what to set these values to? And if issues like this / #35290 keep coming up, isn't trial and error a terrible user experience?

So I'm thinking it could be a long pole to properly design / verify how this should work. In the meantime decoupling the operator from it at least would mean that users don't have to further adjust their statup probe.

**Comment by vmuzikar** [comment]
The way I understood it is that liquibase creates multiple transactions so the timeout is for "migration step", not the whole migration. So it's true that makes it less likely users hit that than the probe timeout. But I'm not sure what each transaction includes (I can check tomorrow), if failing a transaction keeps the DB in a consistent state (e.g. if each transaction covers changes for a single KC version) – AFAIK it does not. I imagine we might need to extend the transaction to cover the whole migration. Trial and error is indeed a terrible UX but IMHO better than risking breaking the DB.

**Comment by sschu** [comment]
@vmuzikar 
>     * made sure the migration is atomic as a whole and if it timeouts, it doesn't leave the DB in an inconsistent state.

AFAIK this is not possible since schema changes (e.g. adding columns) don't run in a transaction.

**Comment by slaskawi** [comment]
> I was trying to keep the long running operation associated with the startup, but having it associated with readiness would work better

@shawkins This sounds good assuming we can use the same criteria for unplugging the Keycloak instance from a Service regardless if it's migration or normal operation. I just want to ensure we thought about it from this point of view. 

>  So I see the logic reordering only as something nice to have as what we truly need to handle is a configurable "migration timeout".

@vmuzikar I think we can reuse the ticket I pointed out earlier: https://github.com/keycloak/keycloak/issues/26889

@shawkins @vmuzikar How do we want Keycloak to behave when the migration fails? Since this is an unrecoverable error, I guess we could just crash?

**Comment by ahus1** [comment]
@vmuzikar - when you talk about the transaction timeout, IMHO we should handle this in a follow-up issue. 

For now, I'd like to focus on if we want to have a quick-fix with a longer start-up probe, and a more long-term approach of moving the migration to phase where startup is complete and the Pod is not yet ready. 

If we think that we wouldn't want to stay in the not-yet-ready phase for too long, we could have a check in the liveness probe that the migration is progressing, and eventually fail the liveness probe when we don't see progress there. Again, this would be a follow-up issue.

**Comment by vmuzikar** [comment]
> we should handle this in a follow-up issue

> quick-fix with a longer start-up probe

+1, let's address it as a quick fix first and deliver a more robust solution (reordering for the readiness probe, timeouts, ...) as a follow-up.

> How do we want Keycloak to behave when the migration fails? Since this is an unrecoverable error, I guess we could just crash?

I think we'll have to discuss what we consider a fail. Hard fail (exception is thrown) – that's clear, crash the pod. But is "progressing too slow" a fail too, as suggested by @ahus1? I'm not sure.

**Comment by ahus1** [comment]
>  But is "progressing too slow" a fail too, as suggested by @ahus1? I'm not sure.

I agree we have more obvious and lower-handing fruits before we get there, and might even not do it all.

**Comment by ahus1** [comment]
FYI, #36308 tracks the increased startup probe timeout for the Operator. It will be part of KC 26.1

**Comment by shawkins** [comment]
> defer the startup logic related to migration, but probably only for kubernetes deployment

@vmuzikar was concerned about introducing a conditional change, so let me elaborate on this more. 

We have an implicit notion of readiness today - we're up once the KeycloakApplication.startup completes .

If we defer / make aynch the migration task, then Keycloak will be reponsive to requests during the migration process. For operator based deployments we're assuming that is ok because the probes will be updated accordingly however if the user has customized the probes, or is not the health probes (much more likely in non-kube deployments) this opens up the potential for serving requests with the database in an inconsistent state around schema changes.

Rather than trying to make when migration is run conditional and / or built upon assumption we should probably make it async all the time, and have a proper internal notion of readiness that would also inhibit regular requests until we're up.

**Comment by stianst** [comment]
I don't have time right now to read through this thread, but some considerations:

* Make sure this doesn't end up in a breaking change in minor releases; non-Kube folks rely on the fact that when KC endpoint is ready; the server is ready to serve requests
* Migrations must complete prior to any requests being handled by KC, and we can't rely on folks to always check the probes to see if KC is ready to serve requests or not

**Comment by vmuzikar** [comment]
Just a random thought. Maybe we could "artificially" reject requests (via some filter maybe) based on the readiness checks? This would allow us reordering startup so migrations is run later but still making sure KC is not listening before it finishes.

**Comment by stianst** [comment]
Migration should be running within a single TX, so unless the TX timeout is drastically increased the migration will fail. At least that's how it used to be; if for whatever reason migration now runs in multiple TX then there are other reasons why KC could be killed during migration, which could cause a partial and broken state; or, if we for some weird reason have bumped up the default TX timeout that would affect all requests, not just migration.

For now I would strongly suggest to roll-back the change in https://github.com/stianst/keycloak/commit/cdebe8c2a384390c45282962e1eda1c3ac56faf9 as this is in my opinion a disruptive change that could have other side-effects; and rather come up with a proper solution and plan for this post 26.1 release.

**Comment by ahus1** [comment]
> if for whatever reason migration now runs in multiple TX

The upgrade has two steps: Liquibase and the Migration on Keycloak's data. The Liquibase migration has always been using multiple transactions, at least one for each changeset, sometimes and also depending on the database additional auto-commits on schema changes. The Migration on Keycloak's data and also updating the version in the migration table is still one transaction. No changes here.

**Comment by vmuzikar** [comment]
> So we might weave that into a decision to run the migration as a Job or with a prolonged startup/readiness probe timeout.

I don't think the migration timeouts is a problem specific to the Operator deployment, nor that the Operator should have any logic for it.

**Comment by pedroigor** [comment]
Would be possible to open a temporary HTTP socket (e.g.: using Vert.x) while the (quarkus application) does not start? If so, we could then serve the liveness probe requests returning a very simple payload/status.

**Comment by slaskawi** [comment]
I'm a bit lost with this thread so let me try to summarize how I understand it:

* Increasing timeout could cause side effects (as per https://github.com/keycloak/keycloak/issues/35261#issuecomment-2582249486) and we pushed figuring this issue out till 26.2.0
* This issue has two layers (as @ahus1 mentioned in https://github.com/keycloak/keycloak/issues/35261#issuecomment-2582314645) - Liquibase and Keycloak data migration. The former runs in multiple transactions, whereas the latter is guarded by the `quarkus.transaction-manager.default-transaction-timeout` and defaults to 300s.
* As @stianst summarized in https://github.com/keycloak/keycloak/issues/35261#issuecomment-2582198865, the Keycloak endpoints should respond when Keycloak is fully operational, which means, all migrations are done.
* In Kubernetes world the above requirement means that the Liveness Probe needs to be up during the migration and the Readiness Probe needs to be down.
* As @pedroigor and @vmuzikar suggested, we may want to introduce a filter that will reject requests when the Keycloak instance is not ready. I would just add that HTTP 503 Service Unavailable seems like a good fit here.

If my summary is correct, there are 3 steps to deliver this functionality:

1. Make the migrations async
2. Design something similar to a filter that will reject all requests during the migrations (or whenever the Readiness Probe is down - we could leverage the same code paths here)
3. Modify the Operator, remove the Startup Probe and potentially adjust timeouts to new defaults (as a consequence of implementing 1 and 2)

Did I capture everything correctly?

**Comment by vmuzikar** [comment]
> migrations async

@pedroigor @slaskawi By async you mean independently of server startup, i.e. the reordering of when migration happens?

> remove the Startup Probe

We had a [discussion](https://github.com/keycloak/keycloak/issues/21111) in the past about the need for startup probe and in the end we [concluded](https://github.com/keycloak/keycloak/pull/21884) in has benefits over just e.g. the liveness probe delay (as it can succeed faster) even when migration is not considered.

**Comment by slaskawi** [comment]
> @pedroigor @slaskawi By async you mean independently of server startup, i.e. the reordering of when migration happens?

@vmuzikar Correct!

> We had a https://github.com/keycloak/keycloak/issues/21111 in the past about the need for startup probe and in the end we https://github.com/keycloak/keycloak/pull/21884 in has benefits over just e.g. the liveness probe delay (as it can succeed faster) even when migration is not considered.

Unless I missed something, with async migrations we wouldn't need it. The Liveness Probe needs to report "up" as soon as possible and keep the "up" state throughout the migrations. Otherwise Kube might kill the Pod. The migrations would be associated only to the Readiness Probe.

**Comment by vmuzikar** [comment]
> Unless I missed something, with async migrations we wouldn't need it. The Liveness Probe needs to report "up" as soon as possible and keep the "up" state throughout the migrations. Otherwise Kube might kill the Pod. The migrations would be associated only to the Readiness Probe.

Migrations aside, the startup probe enables us to account for the server startup time allowing us to use more strict timings  on liveness probe then. The server startup time is especially a bit slower when re-augmentation happens. Possibly, we could  conditionally remove the startup probe just when a custom pre-optimized image is used.

**Comment by ahus1** [comment]
I just found that we don't document how to change the KC database transaction timeout or have an official option. Should I create a separate issue for that? Should we name it `transaction-timeout` as we already have the setting `transaction-xa-enabled`?

**Comment by vmuzikar** [comment]
@ahus1 +1 for having a separate issue for that.

> Should we name it `transaction-timeout`

Sounds good. If we ever decided to have some migration related timeouts, we could use a similar pattern to logging, so something like `transaction-migration-timeout`.

**Comment by ahus1** [comment]
Issue #26889 added and linked to the parent feature.

**Comment by shawkins** [comment]
After more discussion we've decided to simply document the workaround of increasing the startup probe and not persue code changes around the migration / internal readiness at this time.
